File: src/sokoenginepy/tessellation/sokoban_tessellation/sokoban_tessellation.py
from __future__ import annotations

from typing import Dict, Final, Optional, Tuple

from ...atomic_move import AtomicMove
from ...direction import Direction
from ...graph_type import GraphType
from ...utilities import COLUMN, ROW, index_1d, inverted, is_on_board_2d
from ..tessellation_base import TessellationBase


class SokobanTessellation(TessellationBase):
    """
    Board is laid out on squares.

    Direction <-> character mapping:

    ====  =====  ====  ====
    LEFT  RIGHT  UP    DOWN
    ====  =====  ====  ====
    l, L  r, R   u, U  d, D
    ====  =====  ====  ====
    """

    _LEGAL_DIRECTIONS: Final[Tuple[Direction, ...]] = (
        Direction.LEFT,
        Direction.RIGHT,
        Direction.UP,
        Direction.DOWN,
    )

    _CHR_TO_ATOMIC_MOVE: Final[Dict[str, Tuple[Direction, bool]]] = {
        AtomicMove.l: (Direction.LEFT, False),
        AtomicMove.L: (Direction.LEFT, True),
        AtomicMove.r: (Direction.RIGHT, False),
        AtomicMove.R: (Direction.RIGHT, True),
        AtomicMove.u: (Direction.UP, False),
        AtomicMove.U: (Direction.UP, True),
        AtomicMove.d: (Direction.DOWN, False),
        AtomicMove.D: (Direction.DOWN, True),
    }

    _ATOMIC_MOVE_TO_CHR: Final[Dict[Tuple[Direction, bool], str]] = inverted(
        _CHR_TO_ATOMIC_MOVE
    )

    _NEIGHBOR_SHIFT: Final[Dict[Direction, Tuple[int, int]]] = {
        Direction.LEFT: (0, -1),
        Direction.RIGHT: (0, 1),
        Direction.UP: (-1, 0),
        Direction.DOWN: (1, 0),
    }

    def neighbor_position(
        self, position: int, direction: Direction, board_width: int, board_height: int
    ) -> Optional[int]:
        row = ROW(position, board_width)
        column = COLUMN(position, board_width)
        row_shift, column_shift = self._NEIGHBOR_SHIFT.get(direction, (None, None))

        if row_shift is None or column_shift is None:
            raise ValueError(direction)

        row += row_shift
        column += column_shift

        if is_on_board_2d(column, row, board_width, board_height):
            return index_1d(column, row, board_width)

        return None

    def __str__(self):
        return "sokoban"

File: src/libsokoengine/tessellation/sokoban_tessellation.cpp
#include "sokoban_tessellation.hpp"
#include "atomic_move.hpp"
#include "board_cell.hpp"
#include "direction.hpp"

#include <map>
#include <memory>

using namespace std;

namespace sokoengine {

using namespace implementation;

const Directions &SokobanTessellation::legal_directions() const {
  static const Directions retv = {Direction::LEFT, Direction::RIGHT, Direction::UP,
                                  Direction::DOWN};
  return retv;
}

position_t SokobanTessellation::neighbor_position(position_t position,
                                                  const Direction &direction,
                                                  board_size_t board_width,
                                                  board_size_t board_height) const {
  position_t row = Y(position, board_width), column = X(position, board_width);
  switch (direction) {
  case Direction::LEFT:
    column += -1;
    break;
  case Direction::UP:
    row += -1;
    break;
  case Direction::RIGHT:
    column += 1;
    break;
  case Direction::DOWN:
    row += 1;
    break;
  default:
    throw invalid_argument(
      "Unsupported Direction received in SokobanTessellation neighbor_position!");
  }

  if (ON_BOARD(column, row, board_width, board_height))
    return index_1d(column, row, board_width);
  else
    return MAX_POS + 1;
}

char SokobanTessellation::atomic_move_to_char(const AtomicMove &rv) const {
  switch (rv.direction()) {
  case Direction::LEFT:
    return rv.is_push_or_pull() ? AtomicMove::L : AtomicMove::l;
  case Direction::UP:
    return rv.is_push_or_pull() ? AtomicMove::U : AtomicMove::u;
  case Direction::RIGHT:
    return rv.is_push_or_pull() ? AtomicMove::R : AtomicMove::r;
  case Direction::DOWN:
    return rv.is_push_or_pull() ? AtomicMove::D : AtomicMove::d;
  default:
    throw invalid_argument("Illegal AtomicMove character in SokobanTessellation!");
  }
}

AtomicMove SokobanTessellation::char_to_atomic_move(char rv) const {
  switch (rv) {
  case AtomicMove::l:
    return AtomicMove(Direction::LEFT, false);
  case AtomicMove::L:
    return AtomicMove(Direction::LEFT, true);
  case AtomicMove::u:
    return AtomicMove(Direction::UP, false);
  case AtomicMove::U:
    return AtomicMove(Direction::UP, true);
  case AtomicMove::r:
    return AtomicMove(Direction::RIGHT, false);
  case AtomicMove::R:
    return AtomicMove(Direction::RIGHT, true);
  case AtomicMove::d:
    return AtomicMove(Direction::DOWN, false);
  case AtomicMove::D:
    return AtomicMove(Direction::DOWN, true);
  default:
    throw invalid_argument("Illegal AtomicMove direction in SokobanTessellation!");
  }
}

string SokobanTessellation::repr() const { return "SokobanTessellation()"; }
string SokobanTessellation::str() const { return "sokoban"; }

} // namespace sokoengine
